### PR TITLE
krt: move event processing to queue

### PR DESCRIPTION
**Pull request by howardjohn**
tl;dr stop deadlocks in krt

Currently, KRT handlers are synchronous. This means that if I have a chain of N collections depending on each other, by the time we reach the Nth collection, our call stack will have the entire stack of N previous collections. Each other these will hold a lock while the handlers are called.

This has a few issues. The biggest issue is **it can deadlock**. I believe this is only possible with some specific sequences of dependencies; unfortunately, I do not fully understand the specific scenarios that trigger it to make it easily reproducible. However, I and others have encountered it on at least 4 occasions (with either changes we didn't end up pushing as-is). The root of this comes from the fact we have multiple entrypoints into recomputation (since we have many dependencies, and each will call our update handler). Additionally, when we do this recomputation, we need to register new handlers on new dependencies, which in turn requires a lock -- this is, as far as I know, the real root cause:
image
Here we can see we may be holding locks to process an event from collection 1, then Fetch collection 1 which attempts to acquires a lock on collection 1: **deadlock**.

Note that while this may seem a contrived setup, its only really because its showing a simple case. Imagine I wanted to depend on `Waypoints` and `Pods` -- a seemingly reasonably use case and setup, but can trigger this.

Another plausible benefit of this is allowing krt to better make use of multiple cores, but I haven't verified this.

Benchmarks:
```
$ benchstat /tmp/old /tmp/new
name                   old time/op    new time/op    delta
Controllers/krt-16        107ms ±11%     100ms ± 5%   ~     (p=0.095 n=5+5)
Controllers/legacy-16    94.4ms ± 9%   100.5ms ± 7%   ~     (p=0.095 n=5+5)

name                   old alloc/op   new alloc/op   delta
Controllers/krt-16       63.4MB ± 0%    63.3MB ± 0%   ~     (p=0.222 n=5+5)
Controllers/legacy-16    60.8MB ± 0%    60.9MB ± 0%   ~     (p=0.730 n=4+5)

name                   old allocs/op  new allocs/op  delta
Controllers/krt-16         763k ± 0%      760k ± 0%   ~     (p=0.095 n=5+5)
Controllers/legacy-16      719k ± 0%      719k ± 0%   ~     (p=0.690 n=5+5)
```

Running all ambient tests:
```
3773 runs so far, 0 failures (100% pass rate). 1.355801297s avg, 3.75083996s max, 1.150561711s min
```

---

## Implementation comments

The core of this change is in the manyCollection implementation. We lift heavily from the Kubernetes informer model here, even pulling their processorListener almost directly. The primary change here is, of course, to implement a queue and to push all incoming events onto this queue for processing. However, a second major change is required too: implementing a distinct queue for each handler.

A few tests have been changed to not make assumptions about invalid ordering assumptions; these tests had a setup like 

```mermaid
graph TD
  a-->b
  c-->b
```

And assumed when `c` got an even it meant `a` did as well; this assumption no longer holds as each handler is independent.

Additionally, because of the async nature, computations may continue to occur after tests have complete but before things are shutdown. This causes race conditions when we mutate global state for feature flags. To handle this, the ambient code has been rewritten to avoid dependencies on mutable global state.

### CODE REVIEW

**Comment by howardjohn** [comment]
Tests are failing due to us not ever actually killing the stopCh (both with and without this change!) and that meaning we leak goroutines now but not before. Will work on a fix, it doesn't impact the main changes as its a bug in the usage of krt not krt itself.

File: pkg/kube/controllers/queue.go [hunk 1]
```diff
@@ -79,6 +80,13 @@ func WithGenericReconciler(f func(key any) error) func(q *Queue) {
 	}
 }
 
+// WithSyncedCallback registers a callback to run when the queue has synced
```
Comment (howardjohn) [inline pkg/kube/controllers/queue.go:83]: remove this, not used

File: pkg/kube/krt/processor.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+// function sends the given notification to `addCh`.  One goroutine
+// runs `pop()`, which pumps notifications from `addCh` to `nextCh`
+// using storage in the ring buffer while `nextCh` is not keeping up.
+// Another goroutine runs `run()`, which receives notifications from
+// `nextCh` and synchronously invokes the appropriate handler method.
+//
+// processorListener also keeps track of the adjusted requested resync
+// period of the listener.
+type processorListener[O any] struct {
+	nextCh chan any
```
Comment (ilrudie) [inline pkg/kube/krt/processor.go:113]: why `chan any`? is there any case where it would not be an `eventSet[O]` that I've just missed?

File: pkg/kube/krt/processor.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+	event           []Event[O]
+	isInInitialList bool
+}
+
+func (p *processorListener[O]) send(event []Event[O], isInInitialList bool) {
+	if isInInitialList {
+		// The initial sync had no items to process, so we would never call syncTracker.Finish().
+		// Just directly mark it as done here
+		if len(event) == 0 {
+			close(p.syncTracker.synced)
```
Comment (ilrudie) [inline pkg/kube/krt/processor.go:154]: This may be straight from upstream and we just keep it as if for that reason but rather than this close which is functional but seems out of place. Should we just have a receiver function for syncTracker like `MarkFinished()` which can handle this?

File: pkg/kube/krt/processor.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"sync"
+	"sync/atomic"
```
Comment (ilrudie) [inline pkg/kube/krt/processor.go:19]: why sync/atomic here rather than Uber's atomic (which seems to be the more commonly used for Istio)?

**Review by ilrudie — COMMENTED** [review]
Initial look at the KRT changes. Will marinade on those a sec and then review the controller updates.

File: pkg/kube/krt/processor.go (hunk 4 above)
Comment (howardjohn) [reply pkg/kube/krt/processor.go:19]: This code is forked from k8s client-go

Comment (ilrudie) [reply pkg/kube/krt/processor.go:19]: makes sense, I figured it might be worth asking even though that was the likely answer

File: pkg/kube/krt/collection.go [hunk 5]
```diff
@@ -350,14 +345,10 @@ func (h *manyCollection[I, O]) onPrimaryInputEventLocked(items []Event[I]) {
 	if len(events) == 0 {
 		return
 	}
-	handlers := h.eventHandlers.Get()
-
 	if h.log.DebugEnabled() {
```
Comment (stevenctl) [inline pkg/kube/krt/collection.go:348]: Why do we have to guard this log? Having it as a `Debugf` isn't enough? Not new in this PR, but just curious

Comment (howardjohn) [reply pkg/kube/krt/collection.go:348]: The general reason is .WithLabels and sometimes the arguments to it or the log line (`len` here is cheap but imagine it was expensive) have some cost. I don't recall for sure if this initial was more costly or if I had previously decided any `WithLabels` was too costly to run on hot paths

File: pkg/kube/krt/processor.go (hunk 2 above)
Comment (howardjohn) [reply pkg/kube/krt/processor.go:113]: It was for some silly reason, plausibly not needed. I have already blocked it from my memory, will rethink it

**Comment by howardjohn** [comment]
> Tests are failing due to us not ever actually killing the stopCh (both with and without this change!) and that meaning we leak goroutines now but not before. Will work on a fix, it doesn't impact the main changes as its a bug in the usage of krt not krt itself.

Ok, new commit fixes this. Not bad actually, we should have done it regardless

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go [hunk 6]
```diff
@@ -125,10 +133,22 @@ type Options struct {
...
 	LookupNetworkGateways LookupNetworkGateways
 	StatusNotifier        *activenotifier.ActiveNotifier
+	Flags                 FeatureFlags
 
 	Debugger *krt.DebugHandler
 }
 
+// KrtOptions is a small wrapper around KRT options to make it easy to provide a common set of options to all collections
+// without excessive duplication.
+type KrtOptions struct {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:143]: maybe move to the krt pkg, so that all callers can use it

File: pilot/pkg/serviceregistry/kube/controller/ambient/policies.go [hunk 7]
```diff
@@ -207,26 +207,31 @@ func PolicyCollections(
...
 		DefaultPolicy.AsCollection(),
 		ImplicitWaypointPolicies,
-	}, krt.WithName("Policies"))
+	}, opts.WithName("Policies")...)
 	return AuthzDerivedPolicies, Policies
 }
 
-func implicitWaypointPolicyName(waypoint *Waypoint) string {
-	if !features.DefaultAllowFromWaypoint || waypoint == nil || len(waypoint.ServiceAccounts) == 0 {
+func implicitWaypointPolicyName(flags FeatureFlags, waypoint *Waypoint) string {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/policies.go:214]: prefer a global var rather than passing it into every function

File: pkg/kube/controllers/queue.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/kube/controllers/queue.go:83]: yeah, `initialSync` can be queried

File: pkg/kube/krt/collection.go [hunk 8]
```diff
@@ -226,11 +223,9 @@ func (h *manyCollection[I, O]) index(extract func(o O) []string) kclient.RawInde
...
 // This is called either when I directly changes, or if a secondary dependency changed. In this case, we compute which I's depended
 // on the secondary dependency, and call onPrimaryInputEvent with them
-func (h *manyCollection[I, O]) onPrimaryInputEvent(items []Event[I], lock bool) {
-	if lock {
-		h.recomputeMu.Lock()
-		defer h.recomputeMu.Unlock()
-	}
+func (h *manyCollection[I, O]) onPrimaryInputEvent(items []Event[I]) {
+	h.blockNewEvents.Lock()
+	defer h.blockNewEvents.Unlock()
```
Comment (hzxuzhonghu) [inline pkg/kube/krt/collection.go:228]: why do we need the lock if all these items are readonly

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go (hunk 6 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:143]: krt already provides a functional options pattern, this is a simple syntax helper specific to this package to provide a common set of options all the collections in this package need

**Comment by howardjohn** [comment]
Moving to draft since I want to spend some more time doing scale and reliability tests and exploring some other aspects

File: pkg/kube/krt/collection.go [hunk 9]
```diff
@@ -582,35 +574,27 @@ func (h *manyCollection[I, O]) Register(f func(o Event[O])) Syncer {
...
-		// We handle events in sequence here, so its always synced at this point/
-		return alwaysSynced{}
-	}
-	return channelSyncer{
-		name:   h.collectionName + " handler",
-		synced: h.synced,
+	if !runExistingState {
+		// If we don't to run the initial state this is simple, we just register the handler.
+		return h.eventHandlers.Insert(f, h.Synced(), nil)
+	}
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/kube/krt/collection.go:580]: RegisterBatch is an interface so its pretty painful to split

File: pilot/pkg/serviceregistry/kube/controller/ambient/policies.go (hunk 7 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/policies.go:214]: The reason this is done is to avoid a dependency on **global mutable** state, which triggers race conditions. This is discussed a bit more in the PR description

File: pkg/kube/krt/collection.go (hunk 8 above)
Comment (howardjohn) [reply pkg/kube/krt/collection.go:228]: I've updated the documentation in the code, but the idea is to make it so we can register handlers safely. When we add a have to send them `[initial state, any future updates]`. if an even is happening while we are sending the initial state, things get wonky. This ensures we can properly send these events. If you look at the k8s client-go informer code you will see the exact same mutex used (with the same name)

**Comment by howardjohn** [comment]
> Moving to draft since I want to spend some more time doing scale and reliability tests and exploring some other aspects

Done! added a lot more testing. At scale, I saw this with some future changes performances >10x better (though, most of the benefit is from the other changes)

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go [hunk 10]
```diff
@@ -112,6 +112,14 @@ type index struct {
 	Network LookupNetwork
 	// LookupNetworkGateways provides a function to lookup all the known network gateways in the system.
 	LookupNetworkGateways LookupNetworkGateways
+	Flags                 FeatureFlags
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:115]: As mentioned in the other comment about this, this is global mutable state which triggers a race condition in the tests.

File: pilot/pkg/serviceregistry/kube/controller/ambient/policies.go (hunk 7 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/policies.go:214]: not in the test code which mutates them for specific tests

File: pkg/kube/kclient/delayed.go [hunk 11]
```diff
@@ -36,11 +36,27 @@ type delayedClient[T controllers.ComparableObject] struct {
...
 	started  <-chan struct{}
 }
 
+type delayedHandler struct {
+	cache.ResourceEventHandler
+	hasSynced delayedHandlerRegistration
+}
+
+type delayedHandlerRegistration struct {
+	hasSynced *atomic.Pointer[func() bool]
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/kube/kclient/delayed.go:50]: Yes. This is the delayed informer, the same atomic pattern is used for every struct in it

File: pkg/kube/krt/collection.go (hunk 8 above)
Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: my mistake it's a different name https://github.com/kubernetes/client-go/blob/37045084c2aa82927b0e5ffc752861430fd7e4ab/tools/cache/shared_informer.go#L389

Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:230]: In k8s it is to protect event handler register. But here in istio, it seems to protect items. That's my confusion, items not need to protect when being distribute. If one element being changed while it is being distributed, it doesnot matter. This is the older one, and the new one will trigger another distribute

Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: Here it is also to protect event handler registration. In both cases it's used the same as far as I can tell - it blocks new events from coming in so that we cleanly register a handler?

Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:230]: In istio  event handler registration is not protected with this lock, it is with mu

new events won't come in

Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: L587 is where the handler registration is protected with this lock?

Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:230]: L160 is registering
```
	// Send out all the initial objects to the handler. We will then unlock the new events so it gets the future updates.
	return h.eventHandlers.Insert(f, h.Synced(), events, h.stop)
```

Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: I don't understand. that is line 600 not 160 and it's inside of the lock...

Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:230]: .Insert method has a lock, i think we donot need blockNewEvents, will have a try

Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: Thanks for the commit, much appreciated. While the current set of tests doesn't catch a difference between this and your commit, I worked on a new test that does. It fails without the lock.

Let me show a timeline for why. (t1/t2=thread1 / thread 2, executing concurrently)

* t1: Add(a). Take `mu`, add the object, release `mu`.
* t2: Register. Take `mu`, get all the objects (`[a]`), send `[a]` to event handler. Note: the mu could be released before or after 'send [a]', it doesn't impact things; your PR moved it to release after the event handler.
* t1: continue Add(a) execution. Now we trigger send `[a]` to event handler.

Result: the event handlers gets `[a]` twice.

Here is an example failure locally when I remove the blockNewEvents lock:

```
=== RUN   TestConformance
=== RUN   TestConformance/informer
2024-11-27T15:58:57.509116Z     info    cluster "fake" kube client started
2024-11-27T15:58:57.511587Z     info    krt     Informer[*v1.ConfigMap] synced  owner=Informer[*v1.ConfigMap]
2024-11-27T15:58:57.511610Z     info    controllers     starting        controller=crd watcher
2024-11-27T15:58:57.837297Z     info    controllers     stopped controller=crd watcher
=== RUN   TestConformance/static_list
=== RUN   TestConformance/join
2024-11-27T15:58:58.161624Z     info    krt     Join[krt_test.Named] synced
=== RUN   TestConformance/manyCollection
2024-11-27T15:58:58.481589Z     info    krt     ManyCollection[string,krt_test.Named] synced    owner=ManyCollection[string,krt_test.Named]
    conformance_test.go:209: unexpected events: [add/a/12 add/a/13 add/a/14 add/a/15 add/a/16 add/a/19 add/a/17 add/a/18 add/a/11]
--- FAIL: TestConformance (1.84s)
    --- PASS: TestConformance/informer (0.87s)
    --- PASS: TestConformance/static_list (0.32s)
    --- PASS: TestConformance/join (0.32s)
    --- FAIL: TestConformance/manyCollection (0.32s)
FAIL
```

Here you can see it got duplicate events.

Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:230]: Yes. Isn't this the expected result for eventual model, upstream k8s can also trigger twice with two events on `a`

Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: The Kubernetes informer cannot duplicate events. The eventually consistency doesn't mean you can get multiple events, it just means you may miss intermediate states (so if a pod goes from v1->v2->v3, you may just see v1->v3).

Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:230]: If the client consumes fast enough, definitely it can handle  v1->v2->v3 

And if istio handles  v1->v2->v3, any caveats bring?

File: pkg/kube/kclient/delayed.go [hunk 12]
```diff
@@ -86,14 +102,20 @@ func (s *delayedClient[T]) ListUnfiltered(namespace string, selector klabels.Sel
...
+		return (*c).AddEventHandler(h)
 	}
+	s.hm.Lock()
+	defer s.hm.Unlock()
+
+	hasSynced := delayedHandlerRegistration{hasSynced: new(atomic.Pointer[func() bool])}
+	hasSynced.hasSynced.Store(ptr.Of(s.delayed.HasSynced))
+	s.handlers = append(s.handlers, delayedHandler{
+		ResourceEventHandler: h,
+		hasSynced:            hasSynced,
```
Comment (hzxuzhonghu) [inline pkg/kube/kclient/delayed.go:116]: what is the `hasSynced` of delayedHandler used for? I cannot see the caller

File: pkg/kube/krt/collection.go (hunk 8 above)
Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: I am just state the properties that Kubernetes guarantees. Like in the core, not even in client libraries. That it doesn't guarantee you see every update (though its super rare, and does not matter to Istio at all).

None of this has anything to do with this PR, so sorry if it added confusion. The main point I was trying to make is:
* Informers does **not** ever give duplicate events
* This PR does **not** ever give duplicate events
* The proposal (to drop the lock) **does** give duplicate events, which I consider a bug

File: pkg/kube/kclient/delayed.go (hunk 12 above)
Comment (howardjohn) [reply pkg/kube/kclient/delayed.go:116]: L54

File: pkg/kube/krt/collection.go (hunk 8 above)
Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:230]: By duplicate event, assume you mean a handler can receive v1->v1->v2->v3. 

Is there a test can verify it? I cannot understand this possibility

Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: Yes I added the test https://github.com/istio/istio/pull/53994/commits/2ad9028412ae9cd010eea59abf851eff427f426c. It fails with the lock removed (inconsistently since it's a race). I showed a failure in this comment https://github.com/istio/istio/pull/53994#discussion_r1860927390

Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:230]: Weird, i run go race test 30 times, all passed

Comment (howardjohn) [reply pkg/kube/krt/collection.go:230]: 30 times is not often, took me 500

**Comment by hzxuzhonghu** [comment]
@howardjohn move here running with pr 500 times, met timeout

```
2024-12-03T07:19:17.581082Z     debug   krt     sync complete   name=tracker time=1.227µs
2024-12-03T07:19:17.581210Z     debug   krt     sync complete   name=tracker time=62.156µs
panic: test timed out after 10m0s
        running tests:
                TestConformance (1s)
                TestConformance/join (0s)

goroutine 62787 [running]:
testing.(*M).startAlarm.func1()
        /usr/local/go/src/testing/testing.go:2373 +0x265
created by time.goFunc
        /usr/local/go/src/time/sleep.go:215 +0x45

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000de71e0, {0x49848c9, 0xf}, 0x4b09fa8)
        /usr/local/go/src/testing/testing.go:1751 +0x851
testing.runTests.func1(0xc000de71e0)
        /usr/local/go/src/testing/testing.go:2168 +0x86
testing.tRunner(0xc000de71e0, 0xc000091a58)
        /usr/local/go/src/testing/testing.go:1690 +0x227
testing.runTests(0xc000012348, {0x6f501e0, 0x11, 0x11}, {0xc000091b00?, 0xc000091b60?, 0x6f797a0?})
        /usr/local/go/src/testing/testing.go:2166 +0x8bf
testing.(*M).Run(0xc000a043c0)
        /usr/local/go/src/testing/testing.go:2034 +0xf18
istio.io/istio/tests/util/leak.CheckMain({0x4ecc620, 0xc000a043c0})
        /root/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x36
istio.io/istio/pkg/kube/krt.TestMain(...)
        /root/go/src/istio.io/istio/pkg/kube/krt/leak_test.go:25
main.main()
        _testmain.go:83 +0x17d
```

**Comment by howardjohn** [comment]
Likely you are using `-count`? I would recommend against that and instead follow https://github.com/istio/istio/wiki/Test-Flakes#reproducing-test-flakes. Regardless it seems we have gotten quite a bit off topic here -- do you still have concerns of the mutex, given I have provided a test case that only passes with the mutex?

**Comment by hzxuzhonghu** [comment]
@howardjohn sorry for previous mistake, https://github.com/hzxuzhonghu/istio/commit/ee72be79954a146d1e8dcc93c9123e59e3b2375b the patch is a little different from before it protects eventHandlers insert and distribute in `onPrimaryInputEvent`

TestConformance does not fail running 1000 times.

```
14m25s: 991 runs so far, 0 failures
14m30s: 997 runs so far, 0 failures
14m35s: 1003 runs so far, 0 failures
14m40s: 1009 runs so far, 0 failures
14m45s: 1015 runs so far, 0 failures
14m50s: 1021 runs so far, 0 failures
```

**Review by hzxuzhonghu — APPROVED** [review]
Thank you for the patience

**Comment by ilrudie** [comment]
🎉
